### PR TITLE
SQL: Implement DATE_FORMAT function

### PR DESCRIPTION
**Pull request by addins**
Implement DATE_FORMAT according to the MySQL spec: https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format by translating to the java.time patterns used in DATETIME_FORMAT.

Relates to #55065

### CODE REVIEW

**Review by astefan — CHANGES_REQUESTED** [review]
@addins thank you for your contribution. As a preliminary review:
1. the scripting (Painless) part is missing (both from tests point of view and the main code part). This is important!
2. I wish you didn't make all those small changes to `docs.csv-spec` by eliminating the white spaces. It makes this PR much larger than it should. If you could revert those changes to this file, this would be a much cleaner PR.

**Comment by addins** [comment]
I'm sorry @astefan I clicked re-review button accidentally before the changes requested done.

I don't know how to implement the scripting part missing, Is there any example I could follow?

**Comment by astefan** [comment]
@addins  it's about the Processor class whose `process` method is used in scripting: https://github.com/elastic/elasticsearch/pull/54832/files#diff-2f663032b1c27e10472d85bea179691ebadb5a537fc7dfd7f0684e51fd03c262R23

You also have [InternalSqlScriptUtils.java](https://github.com/elastic/elasticsearch/pull/54832/files#diff-9c77c4e7773b17218e7f624247fb36f4a0f1a06b240051752c8eb156ae9c56fa) where you need to define the painless method and [sql_whitelist.txt](https://github.com/elastic/elasticsearch/pull/54832/files#diff-3444e0ddd4e581b4c08e1b954df0bf1ca5fc11596bc83da6a882365c31e58d7f) for the same method's fingerprint.

Judging by the code you have so far, you already started looking at the PR I referenced above twice, I guess you need to have a deeper look at it and see what are the missing parts. Also, for tests that are relevant for the painless scripting, [everything](https://github.com/elastic/elasticsearch/pull/54832/files#diff-fef5f53668764446bc41eff295ff287e04d05661b63d8b486543a3168ac46ee4) that use a simple condition (WHERE) or an aggregate condition (HAVING) should generate painless scripts.

You are also missing an important part that's [testing the translation of a query ](https://github.com/elastic/elasticsearch/pull/54832/files#diff-4d45e91fef75905ca882b13a487d38063d2857fa39a24293afdce1f278198818) using a `date_format` function.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java [hunk 1]
```diff
@@ -36,18 +36,61 @@
         {"F", "S"},
         {"z", "X"}
     };
+
+    private static final String[][] JAVA_TIME_FORMAT_REPLACEMENTS_FOR_MYSQL = {
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:40]: It's great that you've added some "real world" tests with the format specifiers, but I think it'd be great if you also added a simple one for each one of them too.
For instance `c`, `e` and `H` aren't tested at all anywhere (I didn't go further, there might be some others) and `e` and `D` are wrong. (The latter is revealed by the `DateTimeFormatProcessorTests#testFormatting` test, btw, [running the tests](DateTimeFormatProcessorTests.testFormatting) should help; use `-p x-pack/plugin/sql` to limit the scope of the tests run).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java [hunk 2]
```diff
@@ -36,18 +36,61 @@
...
+        DATE_FORMAT;
 
         private String getJavaPattern(String pattern) {
             if (this == FORMAT) {
-                for (String[] replacement : JAVA_TIME_FORMAT_REPLACEMENTS) {
-                    pattern = pattern.replace(replacement[0], replacement[1]);
-                }
+                pattern = replacePattern(pattern, JAVA_TIME_FORMAT_REPLACEMENTS);
+            } else if (this == DATE_FORMAT) {
+                pattern = replacePattern(pattern, JAVA_TIME_FORMAT_REPLACEMENTS_FOR_MYSQL);
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:86]: This is unfortunately not enough. The reason being that the result of `getJavaPattern()` is later fed directly to `DateTimeFormatter.ofPattern()`, which will interpret any other characters which aren't `DATE_FORMAT` specific, but which are `DateTimeFormatter` specific.

`SELECT DATE_FORMAT('2020-03-04T05:06:07Z'::DATETIME, '%A %c')` will not produce `%A 3` (once `c` is corrected), but `%18367000 3` (well, depending on the momentary `milli-of-day`).

This is a current defect with `FORMAT` [too](https://github.com/elastic/elasticsearch/issues/66560). You don't have to necessarily fix `FORMAT`, but a correct implementation of this function would still be needed.

**Review by bpintea — CHANGES_REQUESTED** [review]
@addins thanks for providing an update to the ticket! I think there's progress in the right direction, but there are a few things that would still need work.

**Comment by astefan** [comment]
@elasticmachine ok to test

**Comment by astefan** [comment]
@elasticmachine update branch

**Comment by astefan** [comment]
@addins there is a failure (as you can see from the CI run):
```
15:08:05   2> REPRODUCE WITH: ./gradlew ':x-pack:plugin:sql:qa:server:security:with-ssl:integTest' --tests "org.elasticsearch.xpack.sql.qa.security.JdbcCsvSpecIT" -Dtests.method="test {command.testShowFunctions}" -Dtests.seed=6A917357AA212170 -Dtests.security.manager=true -Dtests.locale=ar-YE -Dtests.timezone=Europe/Busingen -Druntime.java=11
15:08:05   2> org.junit.ComparisonFailure: Different result for column [name], entry [43] expected:<DATE_[PARSE]> but was:<DATE_[FORMAT]>
```

**Comment by matriv** [comment]
> @addins there is a failure (as you can see from the CI run):
> 
> ```
> 15:08:05   2> REPRODUCE WITH: ./gradlew ':x-pack:plugin:sql:qa:server:security:with-ssl:integTest' --tests "org.elasticsearch.xpack.sql.qa.security.JdbcCsvSpecIT" -Dtests.method="test {command.testShowFunctions}" -Dtests.seed=6A917357AA212170 -Dtests.security.manager=true -Dtests.locale=ar-YE -Dtests.timezone=Europe/Busingen -Druntime.java=11
> 15:08:05   2> org.junit.ComparisonFailure: Different result for column [name], entry [43] expected:<DATE_[PARSE]> but was:<DATE_[FORMAT]>
> ```

@addins You need to add the new function in the SHOW FUNCTIONS results in `docs.csv-spec` and `command.csv-spec`.

Furthermore you are still missing integration tests in `datetime.csv-spec`

**Comment by addins** [comment]
@astefan @matriv thanks for pointing the direction. I tried the command to reproduce the failure but I got this error, could you help me?

```
PS C:\Users\Addin\Documents\GitHub\elasticsearch> .\gradlew ':x-pack:plugin:sql:qa:server:security:with-ssl:integTest' --tests "org.elasticsearch.xpack.sql.qa.security.JdbcCsvSpecIT" -D tests.method="test {command.testShowFunctions}" -D tests.seed=6A917357AA212170 -D tests.security.manager=true -D tests.locale=ar-YE -D tests.timezone=Europe/Busingen -D runtime.java=11

> Configure project :modules:reindex
Disabling reindex-from-old tests because we can't get the pid file on windows

> Configure project :plugins:repository-hdfs
hdfsFixture unsupported, please set HADOOP_HOME and put HADOOP_HOME\bin in PATH
=======================================
Elasticsearch Build Hamster says Hello!
  Gradle Version        : 6.6.1
  OS Info               : Windows 10 10.0 (amd64)
  Runtime JDK Version   : 11 (JDK)
  Runtime java.home     : C:\Program Files\AdoptOpenJDK\jdk-11.0.4.11-openj9
  Gradle JDK Version    : 15 (AdoptOpenJDK)
  Gradle java.home      : C:\Program Files\AdoptOpenJDK\jdk-15.0.1.9-hotspot
  Random Testing Seed   : 6A917357AA212170
  In FIPS 140 mode      : false
=======================================

> Task :x-pack:plugin:sql:qa:server:security:with-ssl:integTest FAILED

=== Standard output of node `node{:x-pack:plugin:sql:qa:server:security:with-ssl:integTest-0}` ===
╗   ? last 40 non error or warning messages from C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\logs\es.stdout.log ?
╗ [2021-01-06T16:34:30.619164700Z] [BUILD] Configuring custom cluster specific distro directory: C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT
╗ [2021-01-06T16:34:33.050428700Z] [BUILD] Setting up 3 additional config files
╗ [2021-01-06T16:34:33.056429600Z] [BUILD] Copying additional config files from distro [C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\elasticsearch.yml, C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\jvm.options, C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\jvm.options.d, C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\log4j2.properties, C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\roles.yml, C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\role_mapping.yml, C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\users, C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\distro\8.0.0-DEFAULT\config\users_roles]
╗ [2021-01-06T16:34:33.062429200Z] [BUILD] Creating elasticsearch keystore with password set to []
╗ [2021-01-06T16:34:37.080707700Z] [BUILD] Adding 2 keystore settings and 0 keystore files
╗ [2021-01-06T16:34:42.784958600Z] [BUILD] Installing 0modules
╗ [2021-01-06T16:34:42.784958600Z] [BUILD] Setting up 1 users
╗ [2021-01-06T16:34:46.691941500Z] [BUILD] Starting Elasticsearch process
╗ [2021-01-06T16:34:49.706998600Z] [BUILD] Stopping node

=== Standard error of node `node{:x-pack:plugin:sql:qa:server:security:with-ssl:integTest-0}` ===
╗   ? last 40 non error or warning messages from C:\Users\Addin\Documents\GitHub\elasticsearch\x-pack\plugin\sql\qa\server\security\with-ssl\build\testclusters\integTest-0\logs\es.stderr.log ?
╗ Exception in thread "main" java.lang.NullPointerException
╗       at org.elasticsearch.tools.launchers.JvmErgonomics.extractHeapSize(JvmErgonomics.java:157)
╗       at org.elasticsearch.tools.launchers.JvmErgonomics.choose(JvmErgonomics.java:57)
╗       at org.elasticsearch.tools.launchers.JvmOptionsParser.jvmOptions(JvmOptionsParser.java:145)
╗       at org.elasticsearch.tools.launchers.JvmOptionsParser.main(JvmOptionsParser.java:98)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':x-pack:plugin:sql:qa:server:security:with-ssl:integTest'.
> process was found dead while waiting for ports files, node{:x-pack:plugin:sql:qa:server:security:with-ssl:integTest-0}

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 38s
436 actionable tasks: 9 executed, 427 up-to-date
```

**Comment by palesz** [comment]
Instead of using the `qa:server:security:with-ssl`, try the `qa:server:single-node`. It executes the same tests against a single node cluster setup which in your case would be enough to reproduce the error.

Use these commands instead:

JdbcCsvSpecIT has only a single test failure:

```
./gradlew ':x-pack:plugin:sql:qa:server:single-node:integTest' --tests "org.elasticsearch.xpack.sql.qa.single_node.JdbcCsvSpecIT" -Dtests.method="test {command.testShowFunctions}" -Dtests.seed=6A917357AA212170 -Dtests.security.manager=true -Dtests.locale=sr-Latn-BA -Dtests.timezone=Brazil/Acre -Druntime.java=11
```

JdbcDocSpecIT has 3 failures:

```
./gradlew ':x-pack:plugin:sql:qa:server:single-node:integTest' --tests "org.elasticsearch.xpack.sql.qa.single_node.JdbcDocCsvSpecIT"  -Dtests.seed=6A917357AA212170 -Dtests.security.manager=true -Dtests.locale=vi -Dtests.timezone=Canada/Saskatchewan -Druntime.java=11
```

**Comment by addins** [comment]
thanks @palesz , but I still got the same error using those commands

**Comment by matriv** [comment]
@addins Can you try to remove the `-D runtime.java=11` at the end of the command?

**Comment by palesz** [comment]
Hi @addins, if you are still working on this PR, I'd like to ask you to look at how the [tests of the `TO_CHAR` function](https://github.com/elastic/elasticsearch/blob/034926651708cbe8e96f28d10e029e1f3bf1a3a9/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeToCharProcessorTests.java#L33) work. Instead of covering a few cases manually, the tests should assert against actual output of the MySQL `DATE_FORMAT` function using randomized inputs (timestamp, timezone, format strings). This way we can cover more edge cases and catch more potential bugs and stay in-sync with the MySQL implementation.

**Comment by costin** [comment]
Closing this PR as there hasn't been any update to it in the last 3 months.

**Pull request closed without merging**